Implement minimal emulation of TMEM caching

mimimi085181 opened the pull request:
This is a remake of https://github.com/dolphin-emu/dolphin/pull/3749

Full credit goes to phire.

Old message:
"If none of the texture registers have changed and TMEM hasn't been invalidated or changed in other ways, we can blindly reuse the old texture cache entries without rehashing.

Not only does this fix the bloom effect in Spyro: A Hero's Tail (The game abused texture cache) but it will also provide speedups for other games which use the same texture over multiple draw calls, especially when safe texture cache is in use."

Changed the pr per phire's instructions to only return the current texture(s), if none of the texture registers were changed. If any texture register was changed, fall back to the default hashing and rebuilding textures from memory.

iwubcode commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -457,18 +472,23 @@ void TextureCacheBase::BindTextures()
 void TextureCacheBase::UnbindTextures()
 {
-  bound_textures.fill(nullptr);
+  // bound_textures.fill(nullptr);
```
Is there a reason we're keeping this in?

iwubcode commented on Source/Core/VideoCommon/BPStructs.cpp:
```diff
@@ -285,6 +285,8 @@ static void BPWritten(const BPCmd& bp)
     if (g_bRecordFifoData)
       FifoRecorder::GetInstance().UseMemory(addr, tlutXferCount, MemoryUpdate::TMEM);
 
+    TextureCacheBase::InvalidateBindPoint(0xff);
```
What does 0xff mean?  Do we ever have a case where we don't use that value?

phire replied:
Can we replace this with an `InvalidateAllBindPoints()` function or something.

If we are going to do invalidation of individual stages, we will need more precise functions than this one which takes a bitmask. No point exposing the internals when they aren't even used.

iwubcode commented:
Great job @mimimi085181 !  My only compalint about this is that there's all this magic bit masking that I don't really understand.

It'd be great to get some comments / move some of the common bit changes into functions to make the code clearer.

iwubcode commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -457,18 +472,23 @@ void TextureCacheBase::BindTextures()
 TextureCacheBase::TCacheEntry* TextureCacheBase::Load(const u32 stage)
 {
+  // if nothing has changed, just keep the current texture.
+  if ((invalidated_binds & (1 << stage)) == 0)
```
By "nothing" what do we mean?  That the bind point for the stage isn't invalid?

iwubcode commented:
Ran this through fifocomparer using fifoci tests.  OGL and Vulkan look fine (other than the obvious changes in Spyro) but there are many differences for D3D.

Coins are missing in NSMB, and some things are missing in HomebrewFifo and Lesson08.  There are also changes in fortune_street, SSTimeStones3Frame, StarFoxAShadows.

I'm not sure why there is the variation in backends..

phire commented:
> EDIT: In phire's original PR, he states he should probably use a bitset (assuming std bitset?) would that make things any clearer?

The magic bitmask has now lost all value as a single change invalidates all stages.

Just replace it with a bool.

mimimi085181 commented:
I decided to keep the magic bitmask, because i use it now for UnbindTextures, after i removed the other magic bitmask. Hopefully keeping invalidated_binds also helps in the future, if somebody wants to write a more precise tmem cache emulation.

The latest version fixes the "d3d issue". The old version did not check, if bound textures were removed. And even worse, it just used textures that were invalidated. The relevant texture in Spyro was invalidated, because it overlaps an efb copy. Opengl did ignore this and it still used the texture anyways...

Performance for safe texture accuracy is looking good. For the last story on efb2tex it's something like: master+fast hashing: 60 fps, master+safe hashing: 30 fps, pr+safe hashing: 48 fps.

iwubcode commented:
The only game that has changes now is the fortune_street fifo, all the other tests resulted in no differences!  It's unclear to me if fortune street is a positive or negative change.

phire commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -450,6 +456,9 @@ TextureCacheBase::TCacheEntry* TextureCacheBase::ReturnEntry(unsigned int stage,
   GFX_DEBUGGER_PAUSE_AT(NEXT_TEXTURE_CHANGE, true);
 
+  // We need to keep track of invalided textures until they have actually been replaced or re-loaded
+  invalidated_binds &= 0xff ^ (1 << stage);
```
Might want to look into that bitset.

phire commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -464,11 +473,21 @@ void TextureCacheBase::BindTextures()
 void TextureCacheBase::UnbindTextures()
 {
-  bound_textures.fill(nullptr);
+  for (size_t i = 0; i < bound_textures.size(); ++i)
```
`std::for_each`

phire commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -1460,6 +1487,18 @@ TextureCacheBase::InvalidateTexture(TexAddrCache::iterator iter)
 
   TCacheEntry* entry = iter->second;
 
+  for (size_t i = 0; i < bound_textures.size(); ++i)
```
you can simplify this with `std::any_of`

phire commented on Source/Core/VideoCommon/TextureCacheBase.h:
```diff
@@ -39,6 +39,7 @@ class TextureCacheBase
     bool is_efb_copy;
     bool is_custom_tex;
     bool may_have_overlapping_textures = true;
+    bool is_invalidated = false;  // used only for tmem cache emulation
```
Maybe invert this to `is_valid`

I'm not sure. Or something more descriptive... like `tmem_only`

phire commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -1460,6 +1487,18 @@ TextureCacheBase::InvalidateTexture(TexAddrCache::iterator iter)
+    // Spyro: A Hero's Tail is known for using such overwritten textures.
+    if (bound_textures[i] == entry && (invalidated_binds & (1 << i)) == 0)
+    {
+      bound_textures[i]->is_invalidated = true;
```
Um... when do we actually delete these invalidated textures?

phire commented:
This has been superseded by #5718

mimimi085181 commented:
@phire: I used std::any_of and friends where possible. I don't see how i can use it, when i need both, the index and the element of the array

The tmem_only textures now are checked everytime on Cleanup(), if there isn't a valid bind for them anymore, they get removed.

I also made sure that tmem_only textures are not used for textures_by_hash or partial updates, i have overlooked that somehow earlier.

iwubcode commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -76,6 +78,8 @@ TextureCacheBase::TextureCacheBase()
 
   SetHash64Function();
+
+  invalidated_binds = 0xff;
```
Can't you replace this with the function?

iwubcode commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -90,6 +94,8 @@ void TextureCacheBase::Invalidate()
 
   texture_pool.clear();
+
+  invalidated_binds = 0xff;
```
This could be replaced too

iwubcode commented on Source/Core/VideoCommon/TextureCacheBase.h:
```diff
@@ -39,6 +39,7 @@ class TextureCacheBase
     bool is_efb_copy;
     bool is_custom_tex;
     bool may_have_overlapping_textures = true;
+    bool tmem_only = false;
```
Can this get a comment to describe what it's for?

iwubcode commented on Source/Core/VideoCommon/TextureCacheBase.h:
```diff
@@ -125,6 +126,8 @@ class TextureCacheBase
   virtual void DeleteShaders() = 0;
 
   TCacheEntry* Load(const u32 stage);
+  static void InvalidateAllBindPoints() { invalidated_binds = 0xff; }
```
If you made it a bitset and there was a constant saying how many binds there were (total_bind_points) this would be something like

```
invalidate_binds = std::bitset<total_bind_points>().set();
```

iwubcode replied:
Actually, thinking about it more, if you went this route, it might be cleaner to change this to "available_bind_points", then this just becomes:

```
available_bind_points.reset()
```

iwubcode commented on Source/Core/VideoCommon/TextureCacheBase.h:
```diff
@@ -125,6 +126,8 @@ class TextureCacheBase
 
   TCacheEntry* Load(const u32 stage);
+  static void InvalidateAllBindPoints() { invalidated_binds = 0xff; }
+  static bool IsValidBind(u8 i) { return (invalidated_binds & (1 << i)) == 0; }
```
With a bitset this would be:

```
return invalidate_binds.test(i);
```

Sorry just trying to persuade you 😄

phire commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -138,7 +144,11 @@ void TextureCacheBase::Cleanup(int _frameCount)
-    if (iter->second->frameCount == FRAMECOUNT_INVALID)
+    if (iter->second->tmem_only)
+    {
+      iter = InvalidateTexture(iter);
```
You need to check it's still not bound to a slot before invalidating.

mimimi085181 replied:
The pr does this in InvalidateTexture(), because i couldn't write nice and clean code for this. I think it's possible for a texture to be bound to different tmem slots, at least i don't want to trust that there aren't. So i'd need to go through all of them, and check if any of them is bound. This is also the reason why i don't want to store the binding in the texture cache entry. The only way i could come up with was a bool and a loop going through the array, since again i'd need the index and the element to check this.

phire replied:
Ah, right you are. Though it is more than a little weird to have a function called Invalidate texture that only invalidates sometimes.

degasus replied:
Cleanup is called at a random position within the frame (at least with XFB emulation), so this might flush your tmem cache a bit too early. What do you think about delaying this invalidation for one frame? It should be straight forward to set frameCount to last_frame_count + 1 - TEXTURE_KILL_THRESHOLD within the invalidation code.

phire commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -464,11 +477,21 @@ void TextureCacheBase::BindTextures()
 void TextureCacheBase::UnbindTextures()
 {
-  bound_textures.fill(nullptr);
+  for (size_t i = 0; i < bound_textures.size(); ++i)
```
Hmmmm...

This function is actually called from two places... One is `TextureCacheBase::Invalidate()`, where you do indeed want to wipe all the textures, even if they are valid.

The other place is `VertexManagerBase::Flush()`. Arguably you don't need to call it from there at all... as it's going to try and call `TextureCacheBase::Load()` on every single texture anyway.

The one complication is `Flush()` will only call `Load()` on the fist `usedtextures` slots. Instead of Calling UnbindTextures, you might want to pass that variable into `BindTextures()` instead.

mimimi085181 replied:
In `TextureCacheBase::Invalidate()` i'm just going to call `InvalidateAllBindPoints();` before `UnbindTextures();`. Thanks for pointing that out, it wouldn't have deleted all textures.

I don't understand your isssue with the call to `UnbindTextures();` in `VertexManagerBase::Flush()`. It always goes through `UnbindTextures();`, where it only unbinds all the invalidated binds, then it loads all used textures, and then (re)binds all textures that are not invalidated. The only thing i see that is a bit weird is that it would rebind textures that are not invalided, but also are not used. Not sure if skipping that would optimize things, or make things worse, if those textures are used again later, with no invalidation in between.

phire replied:
It's more that the call to `UnbindTextues` is absolutely pointless.

If the textures have changed, then the `Load` call replaces the bind. If the texture hasn't changed then the bind is bound with the same texture.

There is no way to actually end up with null in a bind point, except for `usedtextures`. Which is why I recommend passing `usedtextures` directly into `BindTextures` to gain the same effect.

Right now this PR changes the behavior slightly when `usedtextures` decreases and a bind slot is not invalidated.

mimimi085181 commented:
@iwubcode: I haven't used std::bitset yet, not sure how to. Maybe it's a better solution, maybe there's an even better one somewhere, i'm not really good at this kind of stuff.

@phire: I changed the binding the way you described, at least if i understood you correctly. Now it should (re)bind all used textures on every flush, about the same way master does right now.

riking commented on Source/Core/VideoCommon/TextureCacheBase.h:
```diff
@@ -125,8 +126,9 @@ class TextureCacheBase
-  virtual void BindTextures();
+  static void InvalidateAllBindPoints() { invalidated_binds = 0xff; }
+  static bool IsValidBind(u8 i) { return (invalidated_binds & (1 << i)) == 0; }
+  virtual void BindTexture(u8 i);
```
Why was this method virtual if there's only one implementation?

mimimi085181 replied:
I removed the virtual there, i hope that's correct, not an expert on these details

iwubcode commented:
@mimimi085181 - see [std::bitset](http://en.cppreference.com/w/cpp/utility/bitset/bitset) it has an example.

Functionally they should be the same, it just seems like it'll be a little clearer to me.

Here's what you have:


```
static u8 invalidated_binds;
static void InvalidateAllBindPoints() { invalidated_binds = 0xff; }
static bool IsValidBind(u8 i) { invalidated_binds & (1 << i)) == 0; }

// We need to keep track of invalided textures until they have actually been replaced or re-loaded
invalidated_binds &= 0xff ^ (1 << stage);
```

Here's what it would look like with a std::bitset:

```
static const int total_bind_points = 8;
static std::bitset<total_bind_points> valid_bind_points;
static void InvalidateAllBindPoints() { valid_bind_points.reset(); }
static bool IsValidBindPoint(u8 i) { return valid_bind_points.test(i); }

// We need to keep track of invalided textures until they have actually been replaced or re-loaded
valid_bind_points.set(stage);
```
(untested but think that should work)

Since it's all been abstracted to a function, it might not matter that much, it's up to you.

mimimi085181 commented:
@iwubcode and @phire: The pr is using std::bitset now. Is that better? The pr is not rebased on top of latest master because of the D3D but it has.

mimimi085181 commented:
Yeah, another run of the fifo comparer sounds nice. Then a (last?) look from phire and we can give it over to JMC to try a bunch of random stuff, especially games where we have hacks for the texture cache. I'm not in a hurry, it doesn't matter to me if it takes a few days more or less to get merged.

iwubcode commented:
Did a preliminary run of fifo comparer and noticed the d3d issue seems to be back.  Some games are not loading its textures, ex: FortunateStreetWhiteBox.dff.  OpenGL and Vulkan show the correct results.

mimimi085181 commented:
I'm getting a bunch of warnings on the FortunateStreetWhiteBox.dff(fifoci.dolphin-emu.org/media/dff/FortuneStreetWhiteBox.dff):
38:44:277 XFStructs.cpp:207 W[Video]: XF load exceeds address space: 1058 1 bytes
38:44:277 XFStructs.cpp:207 W[Video]: XF load exceeds address space: 1059 1 bytes
38:44:277 XFStructs.cpp:207 W[Video]: XF load exceeds address space: 105a 1 bytes

Maybe there's a relation to this issue? Opengl and vulkan were ignoring an issue before while d3d didn't accept it.

mimimi085181 commented:
Ok, the issue with FortunateStreetWhiteBox.dff is now fixed. D3D doesn't seem to like it, if Dolphin just binds the used textures and later reuses a texture that wasn't used in-between? I'm not really sure, but once i switched over to (re)binding all validly bound textures after loading all currently used textures, the issue was fixed. There are textures that are validly bound, but are not currently used.

Maybe that should be commented somewhere in the code, but i wouldn't know where and what exactly.

JMC47 commented:
I tested a few games that are known to do tmem shenanigans, like Spyro, Pokemon, Metroid Prime, Tales of Symphonia, etc.

I also tested safe texture cache heavy games like Silent Hill, CastleVania III, among others.  I didn't notice any defects in a go-through.

autofire372 commented:
What does Metroid Prime do that's TMEM-related?

JMC47 commented:
I was told something with the way the map was drawn could do tmem shenanigans.

degasus commented on Data/Sys/GameSettings/G5S.ini:
```diff
@@ -16,7 +16,3 @@ EmulationIssues = Needs efb to Ram for proper lighting.
 # Add action replay cheats here.
-
-[Video_Hacks]
-EFBToTextureEnable = False
```
You might also want to touch EmulationIssues

JosJuice replied:
I can do that in PR #5760.

orbea commented:
@mimimi085181 and @phire I'm not sure if you're familiar, but this PR is implicated in this issue.

https://bugs.dolphin-emu.org/issues/11164
